krm-functions/dnn-fn/main.go

```diff
@@ -0,0 +1,183 @@
+/*
+ Copyright 2023 The Nephio Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package main
+
+import (
+	"os"
+
+	"fmt"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	nephioreqv1alpha1 "github.com/nephio-project/api/nf_requirements/v1alpha1"
+	infrav1alpha1 "github.com/nephio-project/nephio-controller-poc/apis/infra/v1alpha1"
+	kptcondsdk "github.com/nephio-project/nephio/krm-functions/lib/condkptsdk"
+	ipamv1alpha1 "github.com/nokia/k8s-ipam/apis/alloc/ipam/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// var _ fn.Runner = &DnnFn{}
+
+const (
+	// TODO: this should go to a lib somewhere
+	SiteAnnotationKey = "nephio.org/site"
+)
+
+type DnnFn struct {
+	sdk  kptcondsdk.KptCondSDK
+	site string
+}
+
+func Run(rl *fn.ResourceList) (bool, error) {
+	myFn := DnnFn{}
+	var err error
+	myFn.sdk, err = kptcondsdk.New(
+		rl,
+		&kptcondsdk.Config{
+			For: corev1.ObjectReference{
+				APIVersion: nephioreqv1alpha1.GroupVersion.Identifier(),
+				Kind:       nephioreqv1alpha1.DataNetworkKind,
+			},
+			Owns: map[corev1.ObjectReference]kptcondsdk.ResourceKind{
+				{
+					APIVersion: ipamv1alpha1.GroupVersion.Identifier(),
+					Kind:       ipamv1alpha1.IPAllocationKind,
+				}: kptcondsdk.ChildRemote,
+			},
+			Watch: map[corev1.ObjectReference]kptcondsdk.WatchCallbackFn{
+				{
+					APIVersion: infrav1alpha1.GroupVersion.Identifier(),
+					Kind:       "ClusterContext",
```

Comment:
Suggestion: instead of embeded "ClusterContext" string you can use below:

Kind:       reflect.TypeOf(infrav1alpha1.ClusterContext{}).Name(),